File: src/go/283.move-zeroes.go
/*
 * @lc app=leetcode id=283 lang=golang
 *
 * [283] Move Zeroes
Given an array nums, write a function to move all 0's to the end of it while maintaining the relative order of the non-zero elements.

Example:

Input: [0,1,0,3,12]
Output: [1,3,12,0,0]
Note:

You must do this in-place without making a copy of the array.
Minimize the total number of operations.

*/
package main

import "fmt"

// @lc code=start
func moveZeroes(nums []int) {
	length := len(nums)
	if length == 0 {
		return
	}

	count := 0
	for idx, val := range nums {
		if val != 0 {
			tmp := nums[count]
			nums[count] = val
			nums[idx] = tmp
			count++
		}
	}
}

func main() {
	nums := []int{1, 0, 2, 0, 5, 0, 6, 8, 9, 0}
	moveZeroes(nums)
	fmt.Printf("%v\n", nums)
}

// @lc code=end

File: src/c/283.move-zeroes.c
/*
Given an array nums, write a function to move all 0's to the end of it while
maintaining the relative order of the non-zero elements.

Example:

Input: [0,1,0,3,12]
Output: [1,3,12,0,0]
Note:

You must do this in-place without making a copy of the array.
Minimize the total number of operations.
*/
#include <stdio.h>
#include <stdlib.h>
void moveZeroes(int* nums, int numsSize) {
  if (numsSize == 0) return;
  int count = 0;
  for (int i = 0; i < numsSize; i++) {
    if (nums[i] != 0) {
      int tmp = nums[i];
      nums[i] = nums[count];
      nums[count] = tmp;
      count++;
    }
  }
}
void print(int* nums, int numsSize) {
  for (int i = 0; i < numsSize; i++) {
    printf("%d,", nums[i]);
  }
  printf("\n");
}
int main(int argc, char const* argv[]) {
#define SIZE 10
  int nums[SIZE] = {0, 1, 0, 3, 12,0,4,0,5,9};
  moveZeroes(nums, SIZE);
  print(nums, SIZE);
  return 0;
}
